### PR TITLE
Unify interface of vector index & scalar index.

### PR DESCRIPTION
**Pull request by longjiquan**
Signed-off-by: dragondriver <jiquan.long@zilliz.com>
issue: #15867 
/kind improvement

This pr does not only support scalar index in index builder,
but also unify the interface of vector index & scalar index.

### CODE REVIEW

File: internal/util/indexcgowrapper/dataset.go [hunk 1]
```diff
@@ -0,0 +1,103 @@
+package indexcgowrapper
+
+import (
+	"github.com/milvus-io/milvus/internal/proto/schemapb"
+	"github.com/milvus-io/milvus/internal/storage"
+)
+
+const (
+	keyRawArr = "key_raw_arr"
```
Comment (xiaofan-luan) [inline internal/util/indexcgowrapper/dataset.go:9]: what is the core reason we cant to make Data a map here? should it be simply a list or slice?

File: internal/core/src/indexbuilder/ScalarIndexCreator-inl.h [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    return index_->Serialize(config_);
+}
+
+template <typename T>
+inline void
+ScalarIndexCreator<T>::Load(const knowhere::BinarySet& binary_set) {
+    index_->Load(binary_set);
+}
+
+// not sure that the pointer of a golang bool array acts like other types.
```
Comment (xiaofan-luan) [inline internal/core/src/indexbuilder/ScalarIndexCreator-inl.h:52]: So any conclusion?

Comment (xiaofan-luan) [reply internal/core/src/indexbuilder/ScalarIndexCreator-inl.h:52]: if we doubt on it, maybe we could simply convert to bool to 2 bitSets?

File: internal/proxy/task.go [hunk 3]
```diff
@@ -3221,6 +3227,18 @@ func (cit *createIndexTask) PreExecute(ctx context.Context) error {
 		return err
 	}
 
+	// skip params check of non-vector field.
```
[the remark this answers is not in the document]
Comment (DragonDriver) [reply internal/proxy/task.go:3230]: agree.

File: internal/util/indexcgowrapper/index.go [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+var (
+	_ CodecIndex = (*CgoIndex)(nil)
+)
+
+type CgoIndex struct {
+	indexPtr C.CIndex
+	close    bool
+}
+
+// TODO: use proto.Marshal instead of proto.MarshalTextString for better compatibility.
```
[the remark this answers is not in the document]
Comment (DragonDriver) [reply internal/util/indexcgowrapper/index.go:49]: Agree. But it'll change a lot of related tests, and in fact these parameters won't be persistented, so compatibility issue won't be the main reason. I'll change this in a new pr.

File: internal/util/indexcgowrapper/helper.go [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+		ret[i] = C.GoString((*C.char)(datas[i]))
+	}
+
+	return ret, nil
+}
+
+func GetBinarySetValue(cBinarySet C.CBinarySet, key string) ([]byte, error) {
+	cIndexKey := C.CString(key)
+	defer C.free(unsafe.Pointer(cIndexKey))
+	ret := C.GetBinarySetValueSize(cBinarySet, cIndexKey)
```
[the remark this answers is not in the document]
Comment (DragonDriver) [reply internal/util/indexcgowrapper/helper.go:43]: Here, we iterate the binary_set to get all [key : value] pairs, we need to preallocate the memory for value and then we can fill the content of value via memory copy in cgo. If we don't set the size, it'll cause memory segmentation fault.

File: internal/core/src/indexbuilder/IndexCreatorBase.h [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+#pragma once
+
+#include "knowhere/common/Dataset.h"
+#include "knowhere/common/BinarySet.h"
+#include <memory>
+#include <knowhere/index/Index.h>
+
+namespace milvus::indexbuilder {
+class IndexCreatorBase {
```
[the remark this answers is not in the document]
Comment (DragonDriver) [reply internal/core/src/indexbuilder/IndexCreatorBase.h:20]: agree. In fact, there are many details related to scalar index to be completed. I'll refator these detail in a new pr.

File: internal/core/src/indexbuilder/ScalarIndexCreator-inl.h (hunk 2 above)
Comment (DragonDriver) [reply internal/core/src/indexbuilder/ScalarIndexCreator-inl.h:52]: Need more tests to prove this. I think maybe we should do some refactor to our `StructuredIndex` in segcore. `StructuredIndex` is the interface of scalar index, it defined `In`, `NotIn`, `Range`. But in fact for bool type, the index shouldn't have `Range` interface.

File: internal/util/indexcgowrapper/dataset.go (hunk 1 above)
Comment (DragonDriver) [reply internal/util/indexcgowrapper/dataset.go:9]: no other core reason, I just borrowed the design of our c++ interface. interface `Build(Dataset)` correspond to c++ `Build(Dataset)`. In c++ implementations, `Dataset` was defined as a map.

File: internal/core/src/indexbuilder/ScalarIndexCreator-inl.h [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    Helper::ParseFromString(index_params_, std::string(index_params));
+    // TODO: create index according to the params.
+    index_ = std::make_unique<knowhere::scalar::StructuredIndexSort<T>>();
+}
+
+template <typename T>
+inline void
+ScalarIndexCreator<T>::Build(const knowhere::DatasetPtr& dataset) {
+    auto size = dataset->Get<int64_t>(knowhere::meta::ROWS);
+    auto data = dataset->Get<const void*>(knowhere::meta::TENSOR);
```
Comment (xige-16) [inline internal/core/src/indexbuilder/ScalarIndexCreator-inl.h:36]: get tensor data for scalar index is Confusing

Comment (DragonDriver) [reply internal/core/src/indexbuilder/ScalarIndexCreator-inl.h:36]: OK, I'll rename this later.

**Pull request merged**